pkg/credit/ledger.go

```diff
@@ -0,0 +1,139 @@
+package credit
+
+import (
+	"encoding/json"
+	"net/http"
+	"sort"
+	"time"
+)
+
+type LedgerEntryType string
+
+// Used to sort ledger entries by type.
+var entryTypeWeight = map[LedgerEntryType]int{
+	LedgerEntryTypeGrantUsage: 1,
+	LedgerEntryTypeReset:      2,
+	LedgerEntryTypeGrant:      3,
+	LedgerEntryTypeVoid:       4,
+}
+
+// Defines values for LedgerEntryType.
+const (
+	LedgerEntryTypeGrant      LedgerEntryType = "GRANT"
+	LedgerEntryTypeVoid       LedgerEntryType = "VOID"
+	LedgerEntryTypeReset      LedgerEntryType = "RESET"
+	LedgerEntryTypeGrantUsage LedgerEntryType = "GRANT_USAGE"
+)
+
+func (LedgerEntryType) Values() (kinds []string) {
+	for _, s := range []LedgerEntryType{
+		LedgerEntryTypeGrant,
+		LedgerEntryTypeVoid,
+		LedgerEntryTypeReset,
+		LedgerEntryTypeGrantUsage,
+	} {
+		kinds = append(kinds, string(s))
+	}
+	return
+}
+
+// LedgerEntry is a credit ledger entry.
+type LedgerEntry struct {
+	ID        *string         `json:"id,omitempty"`
+	Type      LedgerEntryType `json:"type"`
+	Time      time.Time       `json:"time"`
+	ProductID *string         `json:"productId,omitempty"`
+	Amount    *float64        `json:"amount,omitempty"`
+	From      *time.Time      `json:"from,omitempty"`
+	To        *time.Time      `json:"to,omitempty"`
+}
+
+// Render implements the chi renderer interface.
+func (c LedgerEntry) Render(w http.ResponseWriter, r *http.Request) error {
+	return nil
+}
+
+func NewLedgerEntryList() LedgerEntryList {
+	return LedgerEntryList{
+		list: []LedgerEntry{},
+	}
+}
+
+type LedgerEntryList struct {
+	list []LedgerEntry
+}
+
+func (c LedgerEntryList) Render(w http.ResponseWriter, r *http.Request) error {
+	return nil
+}
+
+func (f LedgerEntryList) GetEntries() []LedgerEntry {
+	list := make([]LedgerEntry, len(f.list))
+	_ = copy(list, f.list)
+
+	// Sort ledger entries by type first
+	sort.Slice(list, func(i, j int) bool {
+		return entryTypeWeight[list[i].Type] < entryTypeWeight[list[j].Type]
+	})
+
+	// Sort ledger entries by time second
```

Comment:
Right now we are sorting by exact timestamp not by second. I don't know if it's just the comment or if we are missing a Truncate(time.Second) at creation.

This also means that the previous sort would only put the grant uses first if they happened at the exact same time.

But we are generating the timestamp on the service's side meaning that the timestamps would only match in case we are running multiple pods from this if the pods are on the same timestamp https://github.com/openmeterio/openmeter/blob/6b32fc5927f663328cf7a0c471b694571ee2086a/internal/credit/postgres_connector/ent/schema/schema.go#L42

(Using the database's NOW() function can be common timesource)

Reply:
Yes, it's just a comment. I removed `second`.